### PR TITLE
[WIP] Lua helpers for plugin development

### PR DESCRIPTION
**Pull request by KillTheMule**
If we want people to write lua plugins, we should give them some tools to ease development. 

A somewhat loose to-do list, just some things I need to remember:

- [ ] Add `v:lua` to `eval.txt`, reference to `if_lua.txt` for the individial functions

- [ ]  Need help:
  - [ ]  Actually make `v:lua` work
  - [x] Decision about proper namespacing (My suggestion: `vim.helpers` for plain lua help function like `split`, and `vim.lua_api.osname` for the function `osname` that gets exported to viml).

 to actually make `v:lua` work

[ci skip]

### CODE REVIEW

**Comment by justinmk** [comment]
Definitely in favor of this, thanks for getting it started. 

I would suggest not to think of it as "plugin help", but rather "augmenting the Nvim stdlib". Up to now the stdlib has been defined by `:help eval.txt`, AKA all of the `f_*` functions in `src/nvim/*.c`.

We want to enhance the Nvim stdlib with Lua functions. These functions are implemented in Lua, but available also from VimL. 

- These new Nvim stdlib (Lua-implemented) functions might be listed in `:help if_lua.txt`, but that's mostly a detail: they are conceptually _sibling-to_ `:help eval.txt`.
- Using Lua functions from VimL should be as "seamless" as possible (TODO/out-of-scope for this PR). 
    - One idea is to populate `v:lua` with wrappers so that VimL arguments/retval can be marshalled to/from Lua easily without the need for `luaeval({escaped-string})`. Similar to how `helpers.funcs` works in the tests.
- `uname` probably doesn't make sense as implemented from `test/helpers.lua` because it deals with the case where `nvim` is not available. Nvim already has `has('win32|unix|mac')`. 
- Some functions will just wrap existing VimL functions, such as `system()`, but with "Lua idioms". 
    - E.g. `v:lua.system()` could return multiple return-values so that the caller doesn't need to save `v:shell_error` in a separate step.
- In other cases like `v:lua.glob()`, it might make sense to have a completely separate implementation which doesn't suffer from the quirks of VimL's `glob()`, and uses Lua patterns instead of VimL wildcards.
- These functions need to be designed with the same mindset as a new VimL function: forward-compatibility, documentation, tests. So we can't just dump a bunch of functions or [penlight](https://github.com/stevedonovan/Penlight) into userspace, because then we have to support it forever.

One weakness of Lua is the lack of canonical "best practices" to draw from. I think https://github.com/leafo/moonscript and https://github.com/leafo/lapis might be good sources for that.

**Comment by KillTheMule** [comment]
Ok, so I'll try to do this with 1-2 functions I found I needed, and we'll see how this works out. About `v:lua`: This basically means just sticking `call luaeval(require(helpers).function...` as a functtion into this table, right? Should this be autogenerated? I'd handcraft it and try to write a test that each such lua function has a viml entry in `v:lua`...

**Comment by KillTheMule** [comment]
How the hell did I close this? I'll reopen with the next commit.

**Comment by KillTheMule** [comment]
Ok, I've made a pretty trivial example to examine how things would look like. I need 2 pointers

- How do I add this function to the lua api? I tried finding out where `vim.api` is defined (or just `vim` as the toplevel table) to no avail. I mean I found `vim.h.generated.h` and know how I can add a C function to the lua api, but now that I'm writing lua code, I'm lost.
- Where would I define `v:lua`? Is that supposed to be a runtime file? Or some other mechanism I don't know of? If it's a runtime file of course the new functions will not be available if something's wrong with VIMRUNTIME (not sure if we need to deal with that, just asking).

**Comment by KillTheMule** [comment]
After a good night's sleep I was able to solve this, but I'm not sure that's the way it should be done. Seems like I can't use `eval('v:lua.osname()')` but have to resort to `eval('lua.osname()')`, and I'm not sure why. Is this the way it should be?

**Comment by KillTheMule** [comment]
With the patient help of bfredl I was able to make it work! Now there's a test to check if all functions in `vim.helpers` (suggested by bfredl) have their counterpart in `v:lua`.

There's snag though: The code isn't loaded automatically. Bfredl suggested putting a file in `runtime/plugin` to achieve that, which would surely be easy. Otoh, some people care for startup times, and I'm not sure if that's what we want then.

Suggestions?

Next step would be to write documentation (I'd say `osname` can stay the way it is), and think about where to ask people which functions they'd want in this spot. If anybody wants to suggest functions to add, please do so :)

**Comment by bfredl** [comment]
> Otoh, some people care for startup times, and I'm not sure if that's what we want then.

Startup time difference could be measured :) Otherwise `vim.helpers` could be imported on first reference using a metatable. But maybe it is not so important: any script could use `local helpers = require('helpers')` anyway, it matters mostly for quick `:lua print(vim.helpers.somefunc)` testing

**Comment by KillTheMule** [comment]
I tried to make `lua` a `VVAR` instead of a global one. I can confirm it exists by manually trying, but seems it's not modifiable, although I've not set any flags. The `type` of it is a dict, but every try to add an entry results in `Vim(function):E689: Can only index a List or Dictionary`. What am I missing here? Docs are kinda scarce about vvars...

Otoh, all the other `vvars` seem different in usage. They seem more like local variables, that make only sense in a certain context (like `v:foldend` only makes sense when you're talking about a particular fold). Are we sure that's what we want? Why aren't we using a global?

**Comment by KillTheMule** [comment]
I've started bugging people about what they'd like to see in a "lua stdlib". @tjdevries,  @nhooyr, @keidax I think you've dabbled in lua code for nvim, care to chime in? Thanks!

**Comment by hkupty** [comment]
Thanks for calling me in.
There are a few patterns I see myself repeating over and over when using lua.

They mostly drill down to those:
#### For table management:
* [`clone`](https://github.com/vigemus/nvimux/blob/master/lua/fns.lua#L7)
* [`keys`](https://github.com/vigemus/iron.nvim/blob/master/lua/iron/util/functional.lua#L12)
* [`values`](https://github.com/vigemus/iron.nvim/blob/master/lua/iron/util/functional.lua#L20)
* [`get`](https://github.com/vigemus/pointer.nvim/blob/4c6d5a9/lua/pointer/utils.lua#L16) (as a safe alternative to direct referencing)
* [`get-in`](https://github.com/vigemus/pointer.nvim/blob/4c6d5a9/lua/pointer/utils.lua#L20) (get in nested tables)

#### For string management:
* [`split`](https://github.com/vigemus/nvimux/blob/master/lua/fns.lua#L29)

#### For functional-ish style:
* [`curry`](https://github.com/vigemus/iron.nvim/blob/eac8f6330/lua/iron/utils.lua#L25)(this is an applied example)

#### For debbuging (now this is somewhat biased and optional though):
* [`inspect`](https://github.com/vigemus/iron.nvim/blob/1802d37/lua/iron/init.lua#L168)(it is a third-party [library](https://luarocks.org/modules/kikito/inspect)).

Even though there may be more, specially on string management, I feel those are the bits  I need to stop thinking about the plugin and start thinking about the language. It'd be great if some of those could be provided.

Also, I'd like to add that it feels great to use lua to develop plugins for neovim *except* for when I need to expose lua to VimL.

Somethings like [mapping](https://github.com/vigemus/pointer.nvim/blob/4c6d5a9/lua/pointer/init.lua#L71) a [lua function](https://github.com/vigemus/pointer.nvim/blob/4c6d5a9/lua/pointer/init.lua#L5) or [exposing](https://github.com/vigemus/nvimux/blob/master/lua/fns.lua#L41) some [commands](https://github.com/vigemus/nvimux/blob/master/lua/nvimux.lua#L249) can be tricky/messy and sometimes [require](https://github.com/vigemus/nvimux/blob/master/lua/nvimux.lua#L242) some [indirection](https://github.com/vigemus/nvimux/blob/master/lua/nvimux.lua#L215) to work properly.

I would love if the API could expose functions to define mappings and/or create commands in a way that those could _natively_ access lua functions/tables. Obviously I understand the complexity of those, but asking wouldn't hurt, right?

Sorry if this got too dense. I'm glad to help.

Cheers,
Henry

**Comment by bfredl** [comment]
I second `inspect.lua`, it is invaluable when debugging functionaltest code. As it has license inline it could just be vendored as `runtime/lua/inspect.lua`.

For mappings and commands, we already have API functions to inspect them, it seems logical to also have API functions for adding mappings and commands (this will also benefit `init.lua` usage, and remote plugins).

To natively support lua references in API would require a bit work, but a quicker workaround could be to have a lua-side wrapper function with the same signature as the API functions, but also accepts lua closures and converts them to the vimL lua commands accessing them from a reference table (as in the linked examples of indirection).

**Comment by hkupty** [comment]
Taking the indirection out of plugins scope can greatly reduce the cost of writing lua plugins. Even if such mapping is still required, it is acceptable and could be further optimized without requiring to migrate the plugins.

**Comment by hkupty** [comment]
One more thing that might be useful would be helpers to shell out (running `jobstart` maybe) with wrappers to lua functions as callback.

Using lua default `io.popen` is ok, but is not async. Trying to do something async with coroutines can make the plugins way more complex. `jobstart` is already in place and could be exposed with some sugarcoating so it feels natural to lua (for instance hiring that indirection layer).

**Comment by bfredl** [comment]
We definitely want `jobstart` natively in lua, it will allow proper binary output as lua string instead of list-of-lines representation of binary output (which everyone confuses with line-based output)

**Comment by tjdevries** [comment]
Seconded what @hkupty has said.

I made some helper functions while working on the LSP branch:

https://github.com/tjdevries/neovim/blob/0d2dd6446c6cf105806b8b4ab11fe8147fbf4c25/runtime/lua/neovim/meta.lua

https://github.com/tjdevries/neovim/blob/0d2dd6446c6cf105806b8b4ab11fe8147fbf4c25/runtime/lua/neovim/util.lua

Not exactly "complete" but should give an idea of a few of the things it would be helpful to have when writing lua plugins.

**Comment by KillTheMule** [comment]
There's basically 2 types of functions we'd need: "Missing functions of lua" that are lua specific, but don't have anything to do with neovim (like `split`), and "Functions to interact with neovim", like a way to add mapping and commands. Should those be separated somehow, or both live under `vim.helpers` and `v:lua`? Logistically I'd like to make a distinction between "Functions that are exported to viml" and "Functions that are not exported to viml", which seems like the same divide pretty much. Thoughts?

I also agree with `inspect`, which basically falls into the first category, but seems pretty essential.

**Comment by hkupty** [comment]
I think we could go with:
`vim.api` - the current api
`vim.api.native` - the variant that wraps the api to accept lua objects as arguments. 
`vim.std` - the standard/core functions 
`vim.std.shared` - the functions that are exposed to viml via `v:lua`

**Comment by timeyyy** [comment]
> I still think we should have 2 categories

**Comment by tjdevries** [comment]
Alternatively, we could also add some native lua without having to do things to the vim api.

`function string:split(...)` will modify all strings to now have split. As well as `function table:copy` or things like that. Not sure if people prefer that over having `vim.std.string.split(str, ...)`

**Comment by bfredl** [comment]
> function string:split(...) will modify all strings to now have split. 

Not sure I like monkeypatching builtin metatables: it might then look a piece of code only uses builtin methods, but then one runs it outside of nvim and it unexpectedly fails.

> Not sure if people prefer that over having vim.std.string.split(str, ...)

For lua util functions we should also consider the standard module pattern of `local split = require('util').split` (we can use a more elaborate name if we are afraid of collisions)

**Comment by hkupty** [comment]
The name is less important than the structure I'd say...

I think the intent is to make sure all those functions were provided by neovim, thus are accessible from the `vim.*` "namespace".

We could have them as `vim.string`, `vim.table` if we agree that adding more levels is irrelevant and actually worse.

I just want to explicitly separate those 4 cases. A flat structure would be:

`vim.api` - current api
`vim.native_api` - api with native wrapper 
`vim.shared` - exposed to viml
`vim.*` - all the helpers

**Comment by KillTheMule** [comment]
As far as I can see, the current api is useable directly from lua, what do you mean by "native wrapper"? Or am I missing something?

**Comment by phodge** [comment]
@KillTheMule: I've opened a PR for supporting `init.lua` as an alternative to `init.vim` - https://github.com/neovim/neovim/pull/8720. Currently it seems like every part of `init.lua` needs to be written as `vim.api.xxx()`. Perhaps there should be some bundled helpers to simplify startup things like setting options?

**Comment by hkupty** [comment]
I added to the pack a group of functions that could take native lua functions as handlers to mappings, jobs or commands.

**Comment by bfredl** [comment]
> I think the intent is to make sure all those functions were provided by neovim, thus are accessible from the vim.* "namespace".

Do we then also want make sure they are available without runtime (we already have the beginning of this mechanism, see `src/nvim/lua/vim.lua` which will be serialized into the binary), or would that be overkill? (For my own purposes, nvim without its runtime is not nvim, but I've heard this concern being raised before)

>  A flat structure would be:

I think we want as flat a structure as is reasonable, but maybe not functions direct in `vim`, because it increases the risk of conflict if we ever get around to implement/port vim 8's lua interface. Also `vim.api` should be reserved for the C API functions that behave as the same RPC API function, better with separate `vim.xxx.` module for API wrappers.

**Comment by hkupty** [comment]
Why wouldn't we then have those under `nvim.xxx`? This way we woudln't conflict with potential vim 8's lua interface. Either way, I agree that no functions should be exposed on the top-level import, be it a flat structure or a nested one.

**Comment by hkupty** [comment]
@phodge I was thinking about what you said on  excessive `vim.api.nvim_...` calls.

A set of helper functions could be in place so the `init.lua` will look like:
```lua
ns.set_mappings{
   ['<C-a>'] = 'something', 
   ['<C-b>'] = 'something else'
} 
```

**Comment by hkupty** [comment]
Mnkkkbhjhighkjn. Jiiuy) yxddw) 9oplln ghygv. Hi898i3nmmmn.   Mlm. Mlj. Bjjjv

**Comment by KillTheMule** [comment]
Ok, namespacing questions pending I've started looking into adding `split`. It's not a [trivial topic](http://lua-users.org/wiki/SplitJoin), but I'd say the following: Let's only take strings as separators (as opposed to regular expressions), and let's reimplement the behavior of [vim's `split`](https://github.com/neovim/neovim/blob/master/src/nvim/testdir/test_listdict.vim#L554). Thoughts on that?

**Comment by tjdevries** [comment]
I think a solid approach would be that any "extra language features" we add (string.split, table.***) match with the vim functionality wherever possible.

This would make porting a vim plugin -> lua plugin much easier and allows us to refer people to already existing documentation.

We could consider doing something like:

```lua
vim.string.split = function(string, pattern, keepempty)
    return vim.api.nvim_call_function('split', {string, pattern, keepempty})
end
```

Not sure if that makes sense to do though, since you'll have packing and unpacking and that depends on vim settings which could be surprising to people. It also wouldn't work if you ran the code outside of the nvim runtime.

**Comment by KillTheMule** [comment]
I think we should implement this behavior in pure lua.

**Comment by bfredl** [comment]
> I think a solid approach would be that any "extra language features" we add (string.split, table.***) match with the vim functionality wherever possible.

Maybe for some things, but certainly not for builtin lua data structures and other primitives (The very point of using lua is that it is _not_ vimL, is it not?). Especially strings in lua are 8-bit clean unlike vimL, so replicating vimL behavior for strings doesn't always make sense, and even when it does, clean lua implementation is better as @KillTheMule says, to avoid truncation or other misrepresentation of binary strings.

**Comment by KillTheMule** [comment]
I've settled on using `vim.helpers` and `vim.lua_api` for now.

I've implemented a split function, mainly from the [lua-users wiki](http://lua-users.org/wiki/SplitJoin) with some adjustments. You can get an iterater from `gsplit`, but also just get an array from `split`. I've special cased `''` as a separator to produce the chars. Other than that, lua patterns are useable. Captures should be useable too, but I need to investigate and test that (firstly though, I need to understand what that even means :laughing: ).

One thing to note is that I'm trying to [catch infinite loops](https://github.com/neovim/neovim/pull/8677/files#diff-173184d6574b9c0fed48bfad30035356R26). You can see from the [test](https://github.com/neovim/neovim/pull/8677/files#diff-2f800fac2c728de640fc28767968e1edR76) it works in this particular case, and I think it's proper in all case, but please have a look at it.

I've not yet tested the `plain` parameter, and I'm not fully sure what it does yet, but just passing it to `find` seems pretty reasonable.

Feedback welcome :)

**Comment by tjdevries** [comment]
I think `vim.string.split` might be nicer, but otherwise it looks good so far to me!

**Comment by KillTheMule** [comment]
Imho the problem with that is that we're blowing up the namespace, we'll probably introduce `vim.string`, `vim.tables`, `vim.functions` and then some. As far as I can see, none of that will contain a lot of functions.

**Comment by bfredl** [comment]
I agree, one namespace for util functions for lua builtin types and other lang enhancements should be enough. The relevant distinction is between these functions on one hand, and functions that expose vim and os/env functionality on the other hand.

**Comment by hkupty** [comment]
It makes sense. However, some functions can be work for both strings and tables. Unless  the function works for both, we'd need to name them`string_split` and `table_split`...

**Comment by bfredl** [comment]
> However, some functions can be work for both strings and tables.  Unless the function works for both, we'd need to name themstring_split and table_split...

I think we can handle this reasonably on a case-by-case basis, if this situation turns up. I don't see the need to _a priori_ introduce an entire sub-library for each type just because of this. Also the _functionality_ should be central, not the what the type of the first input argument happens to be. I would expect `split` and `join` which are two sides of the same coin to be in the _same_ module, even if they have string and array-like table as different input types.

**Comment by KillTheMule** [comment]
I've now added [inspect](https://github.com/kikito/inspect.lua). License is included in the file, which [should be enough](https://github.com/kikito/inspect.lua#installation). Is it enough to refer to the github site for documentation? Personally, I wouldn't want to blow up the doc file with all the possible options and details, since most of the time it simply "does what you want".

**Comment by KillTheMule** [comment]
I also added trim and deepcopy. I'd say for a first shot, `vim.helpers` is done. I'd do some things for `vim.lua_api` next.

**Comment by KillTheMule** [comment]
@hkupty has started working on a possible api for `map` functions, and we'll use his repository to iterate on that a bit. If anyone wants to participate, [head over](https://github.com/Vigemus/tycho.nvim/issues/2).

**Comment by KillTheMule** [comment]
After some iteration over at tycho, we settled for a simple design for `map`, which I've implemented. I've also made it much more complicated by accounting for modes, recursiveness and cmd-ness. Needs tests :)

I've removed the dependency on the extmarks PR by providing namespaces in the lua code, but in a maximally trivial way.

**Comment by hkupty** [comment]
Just pinging to check if this is still alive. I've been on a rush and couldn't be closer to this change.

**Comment by KillTheMule** [comment]
It's somewhat on hiatus, since I don't have much time these days. I'll pick it up later, or if somebody else wants to take over, be my guest :)

File: runtime/doc/if_lua.txt [hunk 1]
```diff
@@ -293,7 +353,76 @@ vim.types						*lua-vim.types*
...
+	Parameters:~
+		{s}     String: String to split
+		{sep}   String: Separator pattern. If empty, split by chars.
+		{plain} Boolean: If false, match {sep} verbatim
+
+	Return:~
+		Iterator of strings, which are the components of {s} after
+		splitting	
+
+vim.helpers.inspect({object}, {options})                   *vim.helpers.inspect*
```
Comment (justinmk) [inline runtime/doc/if_lua.txt]: Most functions should live directly on `vim.`, not a sub-module `vim.helpers`. We own the `vim.*` module; the only "forbidden" names are those used by legacy `if_lua`:

```
vim
:lua for k,v in pairs(vim) do print(k) end

buffer
open
window
lastline
firstline
type
line
eval
dict
beep
list
command
```

**Comment by KillTheMule** [comment]
Superseeded by #9463.

**Pull request closed without merging**